plugins/net_plugin/net_plugin.cpp

```diff
@@ -4493,32 +4497,37 @@ namespace eosio {
          return "invalid peer address";
       }
 
-      std::lock_guard g( connections_mtx );
-      if( find_connection_i( peer_address ) )
-         return "already connected";
+      {
+         std::lock_guard g(connections_mtx);
+         if (find_connection_i(peer_address))
+            return "already connected";
+      }
 
       auto [host, port, type] = split_host_port_type(peer_address);
 
       auto resolver = std::make_shared<tcp::resolver>( my_impl->thread_pool.get_executor() );
 
       resolver->async_resolve(host, port, 
-         [resolver, host = host, port = port, peer_address = peer_address, listen_address = listen_address, this]( const boost::system::error_code& err, const tcp::resolver::results_type& results ) {
+         [this, resolver, host, port, peer_address, listen_address]( const boost::system::error_code& err, const tcp::resolver::results_type& results ) {
             connection_ptr c = std::make_shared<connection>( peer_address, listen_address );
-            c->set_heartbeat_timeout( heartbeat_timeout );
-            std::lock_guard g( connections_mtx );
-            auto [it, inserted] = connections.emplace( connection_detail{
-               .host = peer_address,
-               .c = std::move(c),
-               .ips = results
+            c->strand.post([this, resolver, c, err, results, host, port, peer_address]() {
+               c->set_heartbeat_timeout( heartbeat_timeout );
+               {
+                  std::lock_guard g( connections_mtx );
+                  connections.emplace( connection_detail{
+                     .host = peer_address,
+                     .c = c,
+                  });
+               }
+               if( !err ) {
+                  c->connect( results );
+               } else {
+                  fc_wlog( logger, "Unable to resolve ${host}:${port} ${error}",
+                           ("host", host)("port", port)( "error", err.message() ) );
+                  c->set_state(connection::connection_state::closed);
+                  ++(c->consecutive_immediate_connection_close);
```

Comment:
What does increasing this do? I thought it was going to cause some sort of cool down when reaching `def_max_consecutive_immediate_connection_close` but doesn't seem to matter

Reply:
Need to look into it, but I think the change that broke this re-connect also broke this back off of `consecutive_immediate_connection_close`

Reply:
Hmm, actually this a reason to reuse the connection object. I guess I should maintain the reuse of the connection object for this.

Reply:
It seems like now (on 51fd8fa) after failing to resolve a host `def_max_consecutive_immediate_connection_close` times it just never tries it again. But if the resolve completes and simply unable to connect to the remote host, it will try retry indefinitely. I can't tell if this discrepancy is intentional or not. It makes me wonder if we should _not_ be increasing this counter here for consistency between the two, but realistically if someone has a completely bad hostname it's probably not fixing itself.